First Iterative Code Review covering Four rules of simple design (Level 3)

sunnykath opened the pull request:
# Kata Information

## Kata name:
Tic Tac Toe Kata

## Kata Description Link
Kata Link: https://github.com/MYOB-Technology/General_Developer/blob/main/katas/kata-foundational/foundational-kata-tictactoe.md

# Learning Outcomes

## Passing Tests - (1st rule of simple design)
### (No Levels)
 Is functionally correct, verified through comprehensive automated tests.

## Reveals Intention - (2nd rule of simple design)
### Level 3
Classes, methods and variable names are named to reveal mainly technical concepts with occasional naming revealing specific domain examples. Comments are very rarely used.

## Removing Duplication - (3rd rule of simple design)
### Level 3
Repeatable patterns are identified based on code duplication. Duplicate code patterns are consistently pulled out into supporting methods or classes.

## Keeping things small - (Fewest Elements, 4th rule of simple design)
### Level 3
Classes and methods are broken up into small components with clear responsibilities. Methods are consistently kept small.

## Object Composition
### Level 3
Can implement object composition in coding problems. Consistently applies composition over inheritance in code.

Commit:
Added Skeleton code for all classes

Commit:
Updated README.md

Commit:
Added User Input interface

Commit:
Implemented InputValidatorTests

Commit:
Added a move class to hold move coordinates and the player making the move

Commit:
Updated the regex to be stricter

Commit:
Added BoardTest file and a test to ensure marker is place correctly - failing

Commit:
Implemented basic place marker functionality

Commit:
Added test to check overriding of marker placements

Commit:
Updated place marker to check cell before writing

Commit:
Added test for checking the board is reset and implemented the function

Commit:
Added a test to check if the game status changes when there's 3 in a row horizontally

Commit:
Added a basic implementation for checkGameStatus and checkWon for horizontal wins

Commit:
Added a test to check if game won when 3 in a row vertically

Commit:
Added Implementation in checkWon to check for vertical wins

Commit:
Added test to check for a diagnol win

Commit:
Updated test name to specify the Diagonal

Commit:
Added a forward slash check in checkWon

Commit:
Added a test to check for backward Slash win

Commit:
Updated Check won to also check for backward slash

Commit:
Removed GameStatus Checks and Simplified place marker

Commit:
Imported Game Status checks

Commit:
Imported game win checks into a game rule handler and implemented fail tests to add to pass tests

Commit:
Added a check for making sure we only checking equivalence for player marks and not empty cells

Commit:
Added a vertical false win test

Commit:
Added false tests for Diagonal wins

Commit:
Added a check for player marks before checking equivalence

Commit:
Updated test names to reflect the assert

Commit:
Implemented user input tests for valid and invalid inputs

Commit:
Added in functionality for the get player move to pass the tests

Commit:
Removed player field from the Move class

Commit:
Removed the cell empty check from board and the corresponding test case

Commit:
Added a test for swapping player turns

Commit:
Implemented Swap Player to swap from player 1 to 2 and vice versa

Commit:
Added test to check if player is swapped back to the initial player after 2 swaps

Commit:
Fixed an error in randomly picking the firs player

Commit:
Added a test for draw on a board with empty cells

Commit:
Added a test for checking draw if no empty cells

Commit:
Implemented check for draw function

Commit:
Replaced foreach loop with a LINQ expression

Commit:
Added a constant class to store all the project constants

Commit:
Updated player function to use contstants instead of magic numbers

Commit:
Refactoring variable names

Commit:
Split Up Functions for getting user input

Commit:
Added Test to check for a quit input from the user

Commit:
Implemented the quit option for user input

Commit:
Added More Constants

Commit:
Refactored Input functions for more readability

Commit:
Added a test to check for duplicate moves

Commit:
Moved the user move input to coordinate indexes translation to the UserInputConsole class instead of the board class

Commit:
Implemented the duplicate move function

Commit:
Renamed Game to GameLoop

Commit:
Added a test to ensure a user can quit the game after an invalid move input

Commit:
Updated README.md with the learning goals for this kata

Commit:
Updated GameLoop to GamePlay

Commit:
Refactoring names, namespaces and line spaces

Commit:
Adding the index adjustment constant for translating the user input to the board indexes

Commit:
Added an extra method to output messages to centralise the output

Commit:
Updated tests to not expect message outputs from all functions

Commit:
Added a test to check if GamePlay accepts and acts on a q input

Commit:
Added quit status to the GameStatus enum

Commit:
Added in functionality to check for a quit input

Commit:
Added a test for checking if the board is updated using the player input move

Commit:
Added functionality to update the board when a player inputs a move

Commit:
Refactored test name to include the message output

Commit:
Added a test for multiple move inputs and swapping player

Commit:
Implemented a while loop to keep updating the board with player inputs until a quit command

Commit:
Added a test to check if duplicate moves are rejected

Commit:
Added a duplicate move message constant and functionality to check for duplicate moves in GamePlay.cs

Commit:
Added test to check for won status if a player makes a winning move

Commit:
Added test to check for a draw

Commit:
Add Game Won message as a Constants.cs

Commit:
Updated the set of input for a draw instead of a win

Commit:
Added functionality to check for wins or draws

Commit:
Refactoring and removing unused code

Commit:
Making the player constant a tuple to also store its Mark

Commit:
Added a constant for empty cells

Commit:
Refactoring to remove rider warnings

Commit:
Updated tests to include indexing adjustment

Commit:
Added test for checking if an empty board is being printed properly

Commit:
Added functionality to output an empty board to the user

Commit:
Added a test to check for printing a non-empty board with player marks

Commit:
Refactoring to avoid warnings

Commit:
Added test to see if the board is being printed constantly between moves in gameplay

Commit:
Implemented functionality to print the board between moves

Commit:
Added test for checking if the board is printed after a win

Commit:
Added functionality to print the board after the game ends and refactored test

Commit:
Refactoring

Commit:
Added the play function to program so the game can be played through the console

Commit:
Updated Player value and mark constants

Commit:
Removed Unused interface

Commit:
Added a function to determine the player mark

Commit:
Creating a function to handle end of game status

Commit:
Refactoring code structure for readability and removing duplication

yzAlvin reviewed (commented):
94 commits nice -- I assume that means committing small increments :D

yzAlvin commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+            };
+        
+            // Act
+            board.PlaceMarker(move, 1);
```
using `int` to represent a player is fine, but I think there could be better ways to `reveal intent`

yzAlvin replied:
eg. just by reading this function the only way I found out `1` probably means player 1 is by looking at the definition of `PlaceMarker`

JamesGolding-myob replied:
Agree with Alvin, you have to look at several of the tests to go "Oh, 1 is probably player one and 2 is probably player two"

yzAlvin commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+            // Arrange 
+            var board = new Board();
+            var move = new Move(1 - Constants.IndexingAdjustment, 1 - Constants.IndexingAdjustment);
```
I think there might be an implicit assumption that whenever we create a new `Move`, we must always remember to `- Constants.IndexingAdjustment` for both x and y, is it possible to do this in the constructor so we don't have to remember to manually do it?

yzAlvin replied:
this affects maintainability, if you quit this project and I pick it up, I might not know that I need to remember to add `- Constants.IndexingAdjustment`

yzAlvin commented on TestTicTacToe/GameRulesTest.cs:
```diff
@@ -0,0 +1,219 @@
+        }
+        
+        [Fact]
+        public void Should_Return_False_When_3_In_A_Row_Diagonally_As_Backward_Slash_Of_Different_Players_Marks()
```
I wonder if all these tests for `Assert.False(hasWonGame)` can be combined into a `theory` or something else -- the scenario being tested is that it should return false when there is a 3 in a row but they aren't all one player, the only difference in each case is where the 3 in a row is

yzAlvin replied:
same for checking for draw and checking for win, might be less verbose to / more readable to have 3 theories:
* not won scenario
* won scenario
* draw scenario

JamesGolding-myob replied:
Agree with Alvin - To keep test files small it is good to combine like scenarios in Theories

yzAlvin commented on TicTacToe/Move.cs:
```diff
@@ -0,0 +1,26 @@
+namespace TicTacToe
+{
+    public class Move
```
for objects like this, you may be interested in looking at `Record` types and how they differ from `Class`, it may make the code less verbose but with the same functionality

yzAlvin commented on TicTacToe/UserInputConsole.cs:
```diff
@@ -0,0 +1,103 @@
+            _moveInput = new Move(moveCoords[0] - Constants.IndexingAdjustment, moveCoords[1] - Constants.IndexingAdjustment);
+        }
+        
+        private static string GetPlayerMark(int player)
+        {
+            var mark = player switch
+            {
+                Constants.PlayerOValue => Constants.PlayerOMark,
+                Constants.PlayerXValue => Constants.PlayerXMark,
+                _ => Constants.EmptyCellMark
+            };
+            return mark;
+        }
```
functions like this make me think instead of representing `players` as an int and having contants to know which is which and what `mark` to use, we could instead have a `Player` object with the mark as being part of the object instead of a value outside

JamesGolding-myob replied:
☝️ I have a similar thought, it seems like we need to encapsulate the player with their token, i.e put them together in a class.

yzAlvin reviewed (commented):
4 rules of simple design
- passes tests, good :D 
- reveals intent, yea i reckon level 3, some things could be more clear, eg. Player
- no duplication, also good, I think some things like isColumnWin and isRowWin is sort of like duplication, they are very very similar
- keep things small, good, lots of abstractions, GamePlay is on the verge of being a bit big, also I am a bit iffy on things like _hasGivenUp and similar things being passed up and down

object composition
- only in the GamePlay class but I think also probably good understanding, though I'm not sure if this is the best kata to demonstrate why it is preferred over inheritance (i can't really imagine where we could use inheritance)

JamesGolding-myob commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace TestTicTacToe
+{
+    public class BoardTests
```
I think I would like to see a test for GetBoard(). Even though it is kinda covered in the other tests. It could be good to show that after a board is instantiated it returns an empty board (i.e full of zeros).

JamesGolding-myob commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+    public class BoardTests
+    {
+        [Fact]
+        public void Should_Place_Marker_In_The_Correct_Position_If_Empty_Cell()
```
I think if you are going to have the condition in this test that the marker is placed in and empty position it would be good to show some sad paths, like you have a board that has the top corner a 1 and you try to place a 1 in that same spot.

Looking at your tests I don't know how you are handling that and one great aspect of having tests is they can act like documentation.

sunnykath replied:
This is a simple refactoring issue. 
Initially, this class was validating for duplicate moves, but at a later stage, it was moved out to GameRulesHandler and this test name wasn't updated. Thank you for pointing it out!

JamesGolding-myob commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+            var actualBoardAfterReset = board.GetBoard();
+        
+            // Assert
+            Assert.Equal(expectedBoardBeforeReset, actualBoardBeforeReset);
```
Be careful with having tests with multiple asserts, sometimes they can add confusion when you are trying to find why a test is failing. It also adds more things in your test that you have to maintain. You have tests that cover the PlaceMarker so it should be pretty safe to just assert that the board returns to empty.
I think in terms of revealing intent I would look at changing the name of the test from Should_Reset_The_Board_Back_To_Its_Initial_State to something like Should_Reset_The_Board_Back_To_Empty/Start, to me initial state could mean there are some starting values in there already, rather than it being all 0's.

JamesGolding-myob commented on TestTicTacToe/BoardTests.cs:
```diff
@@ -0,0 +1,65 @@
+        
+            // Act
+            board.PlaceMarker(move, 1);
+            var actualBoard = board.GetBoard();
```
I don't think this variable actualBoard is providing much value so in terms of keeping things small, I would just put board.GetBoard() in the Assert directly.

JamesGolding-myob commented on TestTicTacToe/GamePlayTests.cs:
```diff
@@ -0,0 +1,288 @@
+            var game = new GamePlay();
+            
+            // Act
+            var initialPlayer = game.GetPlayer();
```
I think changing GetPlayer() to get CurrentCurrentPlayer() would be better,
I'm also a bit concerned as to why SwapPlayer is a public method. It looks like it has been done purely for testing but I think it should be tested as part of handleMoveInput()

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+            return Constants.MoveAcceptedMessage;
+        }
+        
+        private static int RandomlyPickPlayerForFirst()
```
Outside the things that you said you were focusing on but consider YAGNI.
Also I think the name could be better maybe RandomlyPickPlayerToGoFirst().
The fact this os a static method makes me think that something else want to use it, so maybe this isn't the best place for this to live - think feature envy as a code smell

JamesGolding-myob commented on TestTicTacToe/GamePlayTests.cs:
```diff
@@ -0,0 +1,288 @@
+        public void Should_End_Game_When_PLayer_Inputs_Quit_Command()
+        {
+            // Arrange
+            var stringWriter = new StringWriter();
```
You are constant initialising new StringWriters in each test, is it possible to refactor these out?
I think there is also a good opportunity to utilise object composition here and then you probably wouldn't need StringWriter and StringReader. Good place to also explore test Doubles

JamesGolding-myob commented on TestTicTacToe/GameRulesTest.cs:
```diff
@@ -0,0 +1,219 @@
+        public void Should_Return_False_When_3_In_A_Row_Horizontally_Of_Different_Players_Marks()
+        {
+            // Arrange
+            var gameLosingBoard = new[,]
```
this board isn't a losing board, the game should still be in progress

JamesGolding-myob commented on TestTicTacToe/GameRulesTest.cs:
```diff
@@ -0,0 +1,219 @@
+            };
+        
+            // Act
+            var hasWonGame = GameRulesHandler.HasWon(gameLosingBoard);
```
I think a better name here might be isGameWon, as the game can only be won when there are 3 of the same token in a row, otherwise it kinda sounds like you lose once three tokens have been placed

JamesGolding-myob commented on TestTicTacToe/GameRulesTest.cs:
```diff
@@ -0,0 +1,219 @@
+        
+        
+        [Fact]
+        public void Should_Return_True_When_GameStatus_Is_Checked_If_3_In_A_Row_Horizontally()
```
I think you need to add in the fact that they are 3 of the same token in row, otherwise it sounds that you win once three of any tokens are placed in a row

JamesGolding-myob commented on TestTicTacToe/GameRulesTest.cs:
```diff
@@ -0,0 +1,219 @@
+                {0, 0, 0},
+                {0, 0, 0}
+            };
+            var duplicateMove = new Move(0, 0);
```
I'm confused about who's responsibility it is to check for duplicate moves, is it the boards or the GameRules. In the board tests you state a position should be empty but it looks like here is where you do that validation. To me this becomes a bit of a problem because if I was to come work on the project and have to update this validation I'd be confused about where the change needs to be made.

JamesGolding-myob commented on TestTicTacToe/InputValidatorTest.cs:
```diff
@@ -0,0 +1,37 @@
+        [InlineData("1,1")]
+        [InlineData("2,2")]
+        [InlineData("1,3")]
+        public void Should_Equal_True_If_Input_Is_Valid(string input)
```
I like that you have used a Theory here, however it is quite difficult to see what a valid input is. It might be better to do all the valid inputs eg 1,1 1,2 1,3 2,1, 2,2 2,3......

JamesGolding-myob commented on TestTicTacToe/UserInputConsoleTests.cs:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        [Fact]
+        public void Should_Print_An_Empty_Board_With_Dots_Representing_Empty_Cells_And_Board_Printing_Message_When_OutputBoard_Is_called()
```
I think this test is too tightly coupled to your implementation details. This makes your tests hard to maintain if a change is introduced to your requirements

JamesGolding-myob commented on TicTacToe/Board.cs:
```diff
@@ -0,0 +1,39 @@
+
+        public void PlaceMarker(Move move, int player)
+        {
+            var (coordX, coordY) = move.GetCoord();
```
I think having getters on your _coordX and _coordY would be better, you could then pass that straight to your board.

JamesGolding-myob commented on TicTacToe/Board.cs:
```diff
@@ -0,0 +1,39 @@
+        public void PlaceMarker(Move move, int player)
+        {
+            var (coordX, coordY) = move.GetCoord();
+            _boardArray[coordX, coordY] = player;
```
I think an improvement here would be to change player to playerToken or playerNumber, it would be a bit close to showing what it actually is rather than a player.

JamesGolding-myob commented on TicTacToe/Constants.cs:
```diff
@@ -0,0 +1,28 @@
+        public const int EmptyCellValue = 0;
+        public const string EmptyCellMark = ".";
+        
+        public const int PlayerXValue = 1;
```
I'm wondering if we have over complicated things a bit by keeping the board as an int[,]. Maybe it would be simpler if it was a string[,]. That way we wouldn't have to keep track of PlayerXValue and PlayerXMark.

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+
+        public void Play()
+        {
+            var uiConsole = new UserInputConsole();
```
possibly outside the scope of things we are looking at but this looks like a dependency so should be up in the constructor.

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+        private GameStatus _gameStatus;
+        private readonly Board _board;
+        private readonly int[,] _boardArray;
+        private bool _givenUp;
```
your UserInputConsole also has a _givenUp or similar named bool, so I am wondering if we need this _givenUp bool in GamePlay

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+        
+        private void UpdateGameStatus()   
+        {
+            _gameStatus = GameRulesHandler.HasWon(_boardArray) ? GameStatus.Won : (GameRulesHandler.HasDrawn(_boardArray) ? GameStatus.Draw : (_givenUp ? GameStatus.Quit : GameStatus.Playing));
```
Really long chained turneries and method calls can be very difficult to follow and find bugs with. I suggest you avoid this kind of thing and look to use the early return pattern and break the different scenarios into different method calls.

JamesGolding-myob commented on TicTacToe/UserInputConsole.cs:
```diff
@@ -0,0 +1,103 @@
+            var inputStrings = playerInput.Split(',');
+            var moveCoords = inputStrings.Select(int.Parse).ToList();
+
+            _moveInput = new Move(moveCoords[0] - Constants.IndexingAdjustment, moveCoords[1] - Constants.IndexingAdjustment);
```
I would say your names are backwards here. moveCoords, looks more like moveInput then moveInput. playerMove is a better name for _moveInput.

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+                    string outputMessage = HandleMoveInput(inputMove);
+                    uiConsole.OutputMessage(outputMessage);
+                }
+                UpdateGameStatus();
```
I think UpdateGameStatus() implies there is a change every time the loop is run, when actually it stays as Playing until, the player quits, there is a win or a draw. So unless a player quits you are going through 5 loops at a minimum before making a change. you also have this _givenUp guy that could be used to exit the loop

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+            _player = _player == Constants.PlayerXValue ? Constants.PlayerOValue : Constants.PlayerXValue;
+        }
+        
+        public GameStatus GetCurrentStatus()
```
I'm not sure GameStatus needs to be public, is there anything else that needs to know the GameStatus?

JamesGolding-myob commented on TicTacToe/GamePlay.cs:
```diff
@@ -0,0 +1,100 @@
+        private int _player;
+        private GameStatus _gameStatus;
+        private readonly Board _board;
+        private readonly int[,] _boardArray;
```
you have both a board and a boardArray which I find confusing. Do you want to be updating the board or the boardArray/

JamesGolding-myob commented on TicTacToe/GameRulesHandler.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace TicTacToe
+{
+    public static class GameRulesHandler
```
I don't think this should be a static class. Would you ever need to have a GameRulesHandler without a Game?

JamesGolding-myob commented on TicTacToe/Move.cs:
```diff
@@ -0,0 +1,26 @@
+
+        public (int, int) GetCoord()
+        {
+            return (_coordX, _coordY);
```
I think this would be better if you made the _coordX and _coordY Properties instead of them just being fields, they could have a public getter and then you wouldn't need GetCoord() and you probably wouldn't need to override the Equals because you can more easily check if the property values are equal to something else.

JamesGolding-myob commented on TicTacToe/UserInputConsole.cs:
```diff
@@ -0,0 +1,103 @@
+    public class UserInputConsole
+    {
+        private Move _moveInput = new Move(-1,-1);
+        private bool _hasGivenUp;
```
I think it is important to be consistent conceptually, so a better name would be _playerHasQuit.

JamesGolding-myob commented on TicTacToe/UserInputConsole.cs:
```diff
@@ -0,0 +1,103 @@
+            return playerInput == Constants.QuitCommand;
+        }
+
+        private void HandleQuitCommand()
```
I don't think a lot of these small one line methods are helping. I think they are just bloating you class.

JamesGolding-myob reviewed (changes requested):
I ran your code and tests on my machine two tests failed: 
TestTicTacToe.UserInputConsoleTests.Should_Accept_The_Quit_Command_After_An_Invalid_Move
TestTicTacToe.GamePlayTests.Should_Print_The_Final_Board_After_The_Game_Ends

also I think your x and y is backwards, x usually represents horizontal an input of 2,1 I would expect to be placed in the middle of the top row (2nd unit of x) but it is placed in the second row first column, you may want to change to using rows and columns. Also typical cartesian coordinates have have x growing from left to right and y grows from bottom to top, so 3,3 would be the top right hand corner not bottom right hand corner.

Didn't see much about object composition - think you need to go back and look into that more.

I think you've bloated a lot of your classes with lots of single line methods when you could just make your fields/variables into properties.

You also seem to have quite a few used using imports.

Commit:
Removing Duplication by joining multiple fact tests into theory

Commit:
Added a GetBoard Test to increase test coverage

Commit:
Renamed the test and also removed redundant asserts

Commit:
Refactoring names to reveal intention

Commit:
removed the function to randomly pick starting player to keep things small and YAGNI

Commit:
Refactor get player to get current player

Commit:
Refactoring and adding comments as per the reviews on the PR

Commit:
using the test constructors to arrange before each test and prevent duplicate code

Commit:
Creating a rules interface and leveraging interfacing to pass in rules & Refactoring tests and classes to be compatible

Commit:
Adding the rules interface

Commit:
Refactoring some formatting

sunnykath commented:
Thank you for all the review and suggestions!

I have refactored quite a few aspects of the code and worked on most of the feedback provided! Especially the ones that were small changes and small things I missed. 

As per some major refactoring and redesigning, I have made notes of them and also added them to the code itself as comments as a reminder for future me. 

@James for the tests, they do fail but when I run them individually or try and debug them, they pass and I'm unsure as to why that was the case. However, since it was a weird problem that seemed like a rabbit hole and I found myself spending too much time on a weird issue. For this kata and the timeframe, I have decided to leave it be (approved by mentor and LB). 

As per the mini-goals for this:-
Object composition: I agree that I haven't displayed it as much in this specific kata. Therefore, I will be leaving this unchanged and maybe come back to it on a later kata. 
4 simple rules of design: After working on your feedback, I believe I have displayed level 3 for all these rules.

If you both could approve the reviews, I'd merge this PR into my main branch. Thank you so much again!
